### PR TITLE
Improve cause information for NLL higher-ranked errors

### PR DESCRIPTION
**Pull request by Aaron1011**
This PR has several interconnected pieces:

1. In some of the NLL region error code, we now pass
   around an `ObligationCause`, instead of just a plain `Span`.
   This gets forwarded into `fulfill_cx.register_predicate_obligation`
   during error reporting.
2. The general InferCtxt error reporting code is extended to
   handle `ObligationCauseCode::BindingObligation`
3. A new enum variant `ConstraintCategory::Predicate` is added.
   We try to avoid using this as the 'best blame constraint' - instead,
   we use it to enhance the `ObligationCause` of the `BlameConstraint`
   that we do end up choosing.

As a result, several NLL error messages now contain the same
"the lifetime requirement is introduced here" message as non-NLL
errors.

Having an `ObligationCause` available will likely prove useful
for future improvements to NLL error messages.

### CODE REVIEW

**Comment by Aaron1011** [comment]
cc @estebank @lqd @matthewjasper @rust-lang/wg-compiler-nll

**Comment by Aaron1011** [comment]
This should probably have a perf run beforehand, now that I think about it.

@bors r-

**Comment by Aaron1011** [comment]
@estebank The regressions are now much smaller. Can you review the latest commit?

File: compiler/rustc_borrowck/src/region_infer/mod.rs [hunk 1]
```diff
@@ -2000,8 +2000,14 @@ impl<'tcx> RegionInferenceContext<'tcx> {
...
-                if let ConstraintCategory::Predicate(def_id, predicate_span) = constraint.category {
-                    Some(ObligationCauseCode::BindingObligation(def_id, predicate_span))
+                if let ConstraintCategory::Predicate(predicate_span) = constraint.category {
+                    // We currentl'y doesn't store the `DefId` in the `ConstraintCategory`
+                    // for perforamnce reasons. The error reporting code used by NLL only
+                    // uses the span, so this doesn't cause any problems at the moment.
+                    Some(ObligationCauseCode::BindingObligation(
+                        CRATE_DEF_ID.to_def_id(),
+                        predicate_span,
+                    ))
```
Comment (estebank) [inline compiler/rustc_borrowck/src/region_infer/mod.rs:2010]: We should create a new ObligationCauseCode variant instead, but it's fine to do this I guess.

File: compiler/rustc_middle/src/mir/query.rs [hunk 2]
```diff
@@ -309,6 +309,9 @@ pub struct ClosureOutlivesRequirement<'tcx> {
     pub category: ConstraintCategory,
 }
 
+// Make sure this enum doesn't unintentionally grow
+rustc_data_structures::static_assert_size!(ConstraintCategory, 12);
+
```
Comment (estebank) [inline compiler/rustc_middle/src/mir/query.rs:314]: 👍

**Review by estebank — APPROVED** [review]

File: compiler/rustc_borrowck/src/region_infer/mod.rs (hunk 1 above)
Comment (Aaron1011) [reply compiler/rustc_borrowck/src/region_infer/mod.rs:2010]: I was trying to mimic the non-NLL code, which also uses `ObligationCauseCode::BindingObligation` for this message:

https://github.com/rust-lang/rust/blob/2b6ed3b675475abc01ce7e68bb75b457f0c85684/compiler/rustc_infer/src/infer/error_reporting/note.rs#L396-L405

**Pull request merged**